crmprtd/bc_hydro/download.py

```diff
@@ -0,0 +1,116 @@
+''' Downloads data from BC hyrdo
+
+BC Hydro posts a rolling window (3 months) observing hourly data
+once a week. Data is in txt files.
+
+It is recommended to run this script once every 3 months to collect
+all available data, supplemented with aggressive monitoring and
+alerting for errors. If the script is run less than 3 months you
+will miss data.
+'''
+
+import pysftp
+from tempfile import TemporaryDirectory
+import logging
+import os
+import sys
+import re
+from zipfile import ZipFile
+from argparse import ArgumentParser
+from datetime import date, timedelta
+
+from crmprtd import logging_args, setup_logging, common_auth_arguments
+
+log = logging.getLogger(__name__)
+
+
+def download(username, gpg_private_key, ftp_server, ftp_dir, start_date,
+             end_date):
+
+    # Connect FTP server and retrieve file
+    try:
+        sftp = pysftp.Connection(ftp_server, username=username,
+                                 private_key=gpg_private_key)
+    except Exception:
+        log.exception("Invalid ftp authentication")
+        sys.exit(1)
+
+    try:
+        with TemporaryDirectory() as tmp_dir:
+            os.chdir(os.path.expanduser(tmp_dir))
```

Comment:
I'd much, much prefer to *not* `os.chdir`. It's a side-effect that is not necessary. Anywhere that you use `os.getcwd()` below, you could just as well have passed `tmp_dir` in and not have changed the directory for the current user. Try to avoid state as much as possible.

Reply:
I remember the reason I did this was mostly because of this argument for [`connection.get`:](https://pysftp.readthedocs.io/en/release_0.2.8/pysftp.html#pysftp.Connection.get_d)
```
localpath (str) – the local path and filename to copy, destination. If not specified, file is copied to local current working directory
```
and if you specify you have to specify an existing file 
`[Errno 2] No such file or directory: '/tmp/tmpcarurmrp/pcic/Test/PCIC_BCHhourly_200914.zip'` 
so I think I would have to make temp files for this which I can do if that is preferable?